File: SiliconLabsApp/ViewControllers/AppSelection/SILAppSelectionViewController.m
//
//  SILAppSelectionViewController.m
//  BlueGecko
//
//  Created by Kamil Czajka on 17/12/2019.
//  Copyright © 2019 SiliconLabs. All rights reserved.
//

#import "SILAppSelectionViewController.h"
#import "SILApp.h"
#import "SILAppSelectionCollectionViewCell.h"
#import "SILDeviceSelectionViewController.h"
#import "SILCentralManagerBuilder.h"
#import "SILHealthThermometerAppViewController.h"
#import "SILRetailBeaconAppViewController.h"
#import <WYPopoverController/WYPopoverController.h>
#import "WYPopoverController+SILHelpers.h"
#import "SILApp+AttributedProfiles.h"
#import "SILCalibrationViewController.h"
#import "SILBluetoothModelManager.h"
#import "SILBluetoothXMLParser.h"
#import "UIImage+SILImages.h"
#import "SILBluetoothBrowserViewController.h"
#import "SILAppSelectionInfoViewController.h"
#import "SILConstants.h"
#import "SILBluetoothBrowser+Constants.h"
#import "UIView+SILShadow.h"
#import "SILConnectedLightingViewController.h"
#if ENABLE_HOMEKIT
#import "SILHomeKitDebugDeviceViewController.h"
#endif

@interface SILAppSelectionViewController () <UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, SILDeviceSelectionViewControllerDelegate, WYPopoverControllerDelegate, SILAppSelectionInfoViewControllerDelegate, SILIOPPopupDelegate>

@property (strong, nonatomic) IBOutlet UIView *allSpace;
@property (strong, nonatomic) WYPopoverController *devicePopoverController;
@property (weak, nonatomic) IBOutlet UIStackView *tilesSpace;
@property (weak, nonatomic) IBOutlet UIView* appsView;
@property (weak, nonatomic) IBOutlet UICollectionView *appCollectionView;
@property (weak, nonatomic) IBOutlet UIImageView *infoImage;
@property (weak, nonatomic) IBOutlet UIView *aboveSpaceAreaView;
@property (weak, nonatomic) IBOutlet UIView *navigationBarView;
@property (weak, nonatomic) IBOutlet UILabel *navigationBarTitleLabel;
@end

@implementation SILAppSelectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupAppCollectionView];
    [self setupBackground];
    [self setupNavigationBar];
    [self addObseverForNotIntentionallyBackFromThermometer];
    _isDisconnectedIntentionally = NO;
    [[SILBluetoothModelManager sharedManager] populateModels];
    [self.allSpace bringSubviewToFront:self.navigationBarView];
    self.appCollectionView.bounces = NO;
}

#pragma mark - Setup View (viewDidLoad)

- (void)setupAppCollectionView {
    [self registerNibs];
    [self setupAppCollectionViewDelegates];
    [self setupAppCollectionViewAppearance];
}

- (void)registerNibs {
      [self.appCollectionView registerNib:[UINib nibWithNibName:NSStringFromClass([SILAppSelectionCollectionViewCell class]) bundle:nil] forCellWithReuseIdentifier:NSStringFromClass([SILAppSelectionCollectionViewCell class])];
}

- (void)setupAppCollectionViewDelegates {
    self.appCollectionView.dataSource = self;
    self.appCollectionView.delegate = self;
}

- (void)setupAppCollectionViewAppearance {
    self.appCollectionView.backgroundColor = [UIColor sil_backgroundColor];
    self.appCollectionView.alwaysBounceVertical = YES;
}

- (void)setupBackground {
    _allSpace.backgroundColor = [UIColor sil_backgroundColor];
    _appsView.backgroundColor = [UIColor sil_backgroundColor];
}

- (void)setupNavigationBar {
    [self setupNavigationBarBackgroundColor];
    [self setupNavigatioBarTitleLabel];
    [self addGestureRecognizerForInfoImage];
}

- (void)setupNavigationBarBackgroundColor {
    _aboveSpaceAreaView.backgroundColor = [UIColor sil_siliconLabsRedColor];
    _navigationBarView.backgroundColor = [UIColor sil_siliconLabsRedColor];
}

- (void)setupNavigatioBarTitleLabel {
    _navigationBarTitleLabel.font = [UIFont robotoMediumWithSize:SILNavigationBarTitleFontSize];
    _navigationBarTitleLabel.textColor = [UIColor sil_backgroundColor];
}

- (void)addGestureRecognizerForInfoImage {
    UITapGestureRecognizer* tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tappedInfoImage:)];
    [_infoImage addGestureRecognizer:tap];
}

- (void)tappedInfoImage:(UIGestureRecognizer *)gestureRecognizer {
        [self presentAppSelectionInfoViewController:YES];
}

#pragma mark - Setup View (viewDidAppear)

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self postClearBluetoothBrowserNotification];
    if (_isDisconnectedIntentionally) {
        [self showThermometerPopover];
    }
}

- (void)postClearBluetoothBrowserNotification {
    [[NSNotificationCenter defaultCenter] postNotificationName:@"DisconnectAllPeripheral" object:self userInfo:nil];
}

- (void)addObseverForNotIntentionallyBackFromThermometer {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(setIsDisconnectedIntentionallyFlag) name:@"NotIntentionallyBackFromThermometer" object:nil];
}

# pragma mark - Applications

- (void)presentDeviceSelectionViewControllerWithApp:(SILApp *)app animated:(BOOL)animated {
    SILDeviceSelectionViewModel *viewModel = [[SILDeviceSelectionViewModel alloc] initWithAppType:app];
    SILDeviceSelectionViewController *selectionViewController = [[SILDeviceSelectionViewController alloc] initWithDeviceSelectionViewModel:viewModel];
    
    selectionViewController.centralManager = [SILCentralManagerBuilder buildCentralManagerWithAppType:app.appType];
    selectionViewController.delegate = self;

    self.devicePopoverController = [WYPopoverController sil_presentCenterPopoverWithContentViewController:selectionViewController
                                                                                 presentingViewController:self
                                                                                                 delegate:self
                                                                                                 animated:YES];
}

- (void)presentCalibrationViewController:(BOOL)animated {
    SILCalibrationViewController *calibrationViewController = [[SILCalibrationViewController alloc] init];

    self.devicePopoverController = [WYPopoverController sil_presentCenterPopoverWithContentViewController:calibrationViewController
                                                                                 presentingViewController:self
                                                                                                 delegate:self
                                                                                                 animated:YES];
}

- (void)showRetailBeaconAppWithApp:(SILApp *)app animated:(BOOL)animated  {
    SILRetailBeaconAppViewController *appViewController = [[SILRetailBeaconAppViewController alloc] init];
    appViewController.app = app;
    [self.navigationController pushViewController:appViewController animated:animated];
}

- (void)showBluetoothBrowserWithApp:(SILApp *)app animated:(BOOL)animated {
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"SILAppBluetoothBrowser" bundle:nil];
    SILBluetoothBrowserViewController* controller = [storyboard instantiateInitialViewController];
    [self.navigationController pushViewController:controller animated:animated];
}

- (void)showAdvertiserWithApp:(SILApp *)app animated:(BOOL)animated {
    SILAdvertiserHomeWireframe* wireframe = [SILAdvertiserHomeWireframe new];
    [self.navigationController pushViewController:wireframe.viewController animated:animated];
    [wireframe releaseViewController];
}

-(void)showIOPEnterDeviceNamePopup: (SILApp *)app animated:(BOOL)animated {
    SILIOPDeviceNamePopup *popupVC = [[SILIOPDeviceNamePopup alloc] init];
    popupVC.delegate = self;
    self.devicePopoverController = [WYPopoverController sil_presentCenterPopoverWithContentViewController:popupVC
                                                                                 presentingViewController:self
                                                                                                 delegate:self
                                                                                                 animated:YES];
}

- (void)showIOPTestList: (NSString * _Nonnull)text {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"SILIOPTest" bundle:nil];
    SILIOPTesterViewController *iopVC = [storyboard instantiateViewControllerWithIdentifier:@"SILIOPBLETesterVC"];
    iopVC.deviceNameToSearch = text;
    [self.navigationController pushViewController:iopVC animated: YES];
}

- (void)showHomeKitDebugWithApp:(SILApp *)app animated:(BOOL)animated {
#if ENABLE_HOMEKIT
    SILHomeKitDebugDeviceViewController *appViewController = [[SILHomeKitDebugDeviceViewController alloc] init];
    appViewController.app = app;
    [self.navigationController pushViewController:appViewController animated:animated];
#endif
}

- (void)showRangeTestWithApp:(SILApp *)app animated:(BOOL)animated {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"SILAppTypeRangeTest" bundle:nil];
    SILRangeTestAppViewController *viewController = [storyboard instantiateInitialViewController];
    [self.navigationController pushViewController:viewController animated:animated];
}

- (void)didSelectApp:(SILApp *)app {
    NSLog(@"didSelectItem: %@", app.title);
    switch (app.appType) {
        case SILAppTypeConnectedLighting:
        case SILAppTypeHealthThermometer:
        case SILAppTypeBlinky:
            [self presentDeviceSelectionViewControllerWithApp:app animated:YES];
            break;
        case SILAppTypeRangeTest:
            [self showRangeTestWithApp:app animated:YES];
            break;
        case SILAppTypeRetailBeacon:
            [self showRetailBeaconAppWithApp:app animated:YES];
            break;
        case SILAppBluetoothBrowser:
            [self showBluetoothBrowserWithApp:app animated:YES];
            break;
        case SILAppTypeAdvertiser:
            [self showAdvertiserWithApp:app animated:YES];
            break;
        case SILAppTypeHomeKitDebug:
            [self showHomeKitDebugWithApp:app animated:YES];
            break;
        case SILAppIopTest:
            [self showIOPEnterDeviceNamePopup:app animated:YES];
        default:
            break;
    }
}

#pragma mark - Info button

- (void)infoTileTapped:(UITapGestureRecognizer*)recognizer {
    [self presentAppSelectionInfoViewController:YES];
}

#pragma mark - Collection View for Applications

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.appsArray.count;
}

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                           cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    SILAppSelectionCollectionViewCell *cell =  [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([SILAppSelectionCollectionViewCell class]) forIndexPath:indexPath];
    
    SILApp *app = self.appsArray[indexPath.row];
    [cell setFieldsInCell:app];
    return cell;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat cellsInRow = 2.0;
    CGFloat minimumLineSpacing = 16.0;
    CGFloat width = floor((self.appCollectionView.frame.size.width - self.appCollectionView.contentInset.left - self.appCollectionView.contentInset.right - self.appCollectionView.alignmentRectInsets.left - self.appCollectionView.alignmentRectInsets.right) / cellsInRow) - minimumLineSpacing;
      
    CGFloat height = 182.0;
    return CGSizeMake(width, height);
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [collectionView deselectItemAtIndexPath:indexPath animated:YES];
    
    SILApp* app = self.appsArray[indexPath.row];
    [self didSelectApp:app];
}

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section {
    return UIEdgeInsetsZero;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 8.0;
}

- (void)deviceSelectionViewController:(SILDeviceSelectionViewController *)viewController didSelectPeripheral:(CBPeripheral *)peripheral {
    [self.devicePopoverController dismissPopoverAnimated:YES completion:^{
        self.devicePopoverController = nil;

        if (viewController.viewModel.app.appType == SILAppTypeHealthThermometer) {
            [self runHealthThermometer:viewController forPeripheral:peripheral];
        } else if (viewController.viewModel.app.appType == SILAppTypeConnectedLighting) {
            SILConnectedLightingViewController *appViewController = [[UIStoryboard storyboardWithName:@"SILAppTypeConnectedLighting" bundle:nil] instantiateInitialViewController];
            appViewController.centralManager = viewController.centralManager;
            appViewController.connectedPeripheral = peripheral;
            [self.navigationController pushViewController:appViewController animated:YES];
        } else if (viewController.viewModel.app.appType == SILAppTypeBlinky) {
            SILAppTypeBlinkyViewController *appViewController = [[UIStoryboard storyboardWithName:@"SILAppTypeBlinky" bundle:nil] instantiateInitialViewController];
            appViewController.centralManager = viewController.centralManager;
            appViewController.connectedPeripheral = peripheral;
            [self.navigationController pushViewController:appViewController animated:YES];
        }
    }];
}

- (void)didDismissDeviceSelectionViewController {
      [self.devicePopoverController dismissPopoverAnimated:YES completion:nil];
}

- (void)runHealthThermometer:(SILDeviceSelectionViewController*)viewController forPeripheral:(CBPeripheral*)peripheral {
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"SILAppTypeHealthThermometer" bundle:nil];
    UIViewController* controller = [storyboard instantiateInitialViewController];
    
    if ([controller isKindOfClass:[SILHealthThermometerAppViewController class]]) {
        SILHealthThermometerAppViewController* healthThermometerController = (SILHealthThermometerAppViewController*) controller;
        healthThermometerController.centralManager = viewController.centralManager;
        healthThermometerController.app = viewController.viewModel.app;
        healthThermometerController.connectedPeripheral = peripheral;
        [self.navigationController pushViewController:healthThermometerController animated:YES];
    }
}

- (void)presentAppSelectionInfoViewController:(BOOL)animated {
    SILAppSelectionInfoViewController *infoViewController = [[SILAppSelectionInfoViewController alloc] init];
    infoViewController.delegate = self;

    self.devicePopoverController = [WYPopoverController sil_presentCenterPopoverWithContentViewController:infoViewController
                                                                                 presentingViewController:self
                                                                                                 delegate:self
                                                                                                 animated:YES];
}

#pragma mark - SILAppSelectionInfoViewControllerDelegate

- (void)didFinishInfoWithAppSelectionInfoViewController:(SILAppSelectionInfoViewController *)infoViewController {
    [self.devicePopoverController dismissPopoverAnimated:YES completion:^{
        self.devicePopoverController = nil;
    }];
}

#pragma mark - WYPopoverControllerDelegate

- (void)popoverControllerDidDismissPopover:(WYPopoverController *)popoverController {
    [self.devicePopoverController dismissPopoverAnimated:YES completion:nil];
    self.devicePopoverController = nil;
}

- (void)setIsDisconnectedIntentionallyFlag {
    _isDisconnectedIntentionally = YES;
}

- (void)showThermometerPopover {
    NSArray<SILApp*>* app = [SILApp demoApps];
    [self presentDeviceSelectionViewControllerWithApp:app[0] animated:YES];
    _isDisconnectedIntentionally = NO;
}

#pragma mark - SILIOPPopupDelegate

- (void)didTappedOKButtonWithDeviceName:(NSString *)text bluetoothState:(BOOL)bluetoothState {
    [self.devicePopoverController dismissPopoverAnimated:YES];
    self.devicePopoverController = nil;
    if (bluetoothState) {
        [self showIOPTestList:text];
    } else {
        SILBluetoothDisabledAlertObjc* bluetoothDisabledAlert = [[SILBluetoothDisabledAlertObjc alloc] initWithBluetoothDisabledAlert:SILBluetoothDisabledAlertInteroperabilityTest];
        
        [self alertWithOKButtonWithTitle:[bluetoothDisabledAlert getTitle]
                                 message:[bluetoothDisabledAlert getMessage]
                              completion:nil];
    }
}

- (void)didTappedCancelButton {
    [self.devicePopoverController dismissPopoverAnimated:YES];
    self.devicePopoverController = nil;
}

@end

File: SiliconLabsApp/ViewControllers/AppSelection/SILAppSelectionViewController.swift
//
//  SILAppSelectionViewController.swift
//  BlueGecko
//
//  Created by Kamil Czajka on 22.3.2021.
//  Copyright © 2021 SiliconLabs. All rights reserved.
//

import Foundation
import SVProgressHUD

@objcMembers
class SILAppSelectionViewController : UIViewController, UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, SILDeviceSelectionViewControllerDelegate, WYPopoverControllerDelegate, SILAppSelectionInfoViewControllerDelegate, SILThunderboardDeviceSelectionViewControllerDelegate {
    var appsArray: [SILApp] = SILApp.demoApps()
    var isDisconnectedIntentionally: Bool = false

    @IBOutlet var allSpace: UIView!
    @IBOutlet weak var tilesSpace: UIStackView!
    @IBOutlet weak var appsView: UIView!
    @IBOutlet weak var appCollectionView: UICollectionView!
    private var devicePopoverController: WYPopoverController?

    private var peripheralManagerSubscription: SILObservableToken?
    private var disposeBag = SILObservableTokenBag()
    private var peripheralManager: SILThroughputPeripheralManager!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setupAppCollectionView()
        self.setupBackground()
        self.addObserverForNonIntentionallyBackFromThermometer()
        self.isDisconnectedIntentionally = false
        SILBluetoothModelManager.shared()?.populateModels()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        if self.isDisconnectedIntentionally {
            self.showThermometerPopover()
        }
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        self.appCollectionView.collectionViewLayout.invalidateLayout()
    }

    private func setupAppCollectionView() {
        self.registerNibs()
        self.setupAppCollectionViewDelegates()
        self.setupAppCollectionViewAppearance()
    }
    
    private func registerNibs() {
        self.appCollectionView.register(UINib(nibName: String(describing: SILAppSelectionCollectionViewCell.self), bundle:nil), forCellWithReuseIdentifier:String(describing: SILAppSelectionCollectionViewCell.self))
    }
    
    private func setupAppCollectionViewDelegates() {
        self.appCollectionView.dataSource = self
        self.appCollectionView.delegate = self
    }
    
    private func setupAppCollectionViewAppearance() {
        self.appCollectionView.backgroundColor = UIColor.sil_background()
        self.appCollectionView.alwaysBounceVertical = true
    }
    
    private func setupBackground() {
        self.allSpace.backgroundColor = UIColor.sil_background()
        self.appsView.backgroundColor = UIColor.sil_background()
    }

    private func addObserverForNonIntentionallyBackFromThermometer() {
        NotificationCenter.default.addObserver(self, selector: #selector(setIsDisconnectedIntentionallyFlag), name: NSNotification.Name(rawValue: "NotIntentionallyBackFromThermometer"), object:nil)
    }
    
    private func presentDeviceSelectionViewController(app: SILApp!, shouldConnectWithPeripheral shouldConnect: Bool = true, animated: Bool,
                                                      filter: DiscoveredPeripheralFilter? = nil) {
        var viewModel: SILDeviceSelectionViewModel?
        if let filter = filter {
            viewModel = SILDeviceSelectionViewModel(appType: app, withFilter: filter)
        }
        else {
            viewModel = SILDeviceSelectionViewModel(appType: app)
        }
        let selectionViewController = SILDeviceSelectionViewController(deviceSelectionViewModel: viewModel!, shouldConnect: shouldConnect)
        selectionViewController.centralManager = SILBrowserConnectionsViewModel.sharedInstance()!.centralManager!
        selectionViewController.delegate = self
        self.devicePopoverController = WYPopoverController.sil_presentCenterPopover(withContentViewController: selectionViewController, presenting: self,
                                                                                    delegate: self, animated: true)
    }
    
    private func presentThunderboardDeviceSelection(app: SILApp!, animated: Bool, filter: ((Device) -> Bool)? = nil) {
        let thunderboardCentralManager = BleManager()
        let interaction = DeviceSelectionInteraction(scanner: thunderboardCentralManager, connector: thunderboardCentralManager, appType: app.appType, filter: filter)
        let selectionViewController = SILThunderboardDeviceSelectionViewController(interaction: interaction, appType: app.appType)
        selectionViewController.delegate = self
        
        self.devicePopoverController = WYPopoverController.sil_presentCenterPopover(withContentViewController: selectionViewController, presenting: self, delegate: self, animated: true)
    }

    private func showRangeTest(app: SILApp!, animated: Bool) {
        let storyboard = UIStoryboard(name: "SILAppTypeRangeTest", bundle: nil)
        if let viewController = storyboard.instantiateInitialViewController() {
            self.navigationController?.pushViewController(viewController, animated: animated)
        }

    }
    
    private func showRSSIGraph(app: SILApp!, animated: Bool) {
        let storyboard = UIStoryboard(name: "SILAppRSSIGraph", bundle: nil)
        if let viewController = storyboard.instantiateInitialViewController() {
            self.navigationController?.pushViewController(viewController, animated: animated)
        }
    }
    
    private func didSelectApp(app: SILApp!) {
        debugPrint("didSelectItem \(String(describing: app.title))")
        switch app.appType {
        case .typeConnectedLighting,
             .typeHealthThermometer:
            self.presentDeviceSelectionViewController(app: app, animated: true)
        
        case .typeBlinky:
            self.presentThunderboardDeviceSelection(app: app, animated: true) { $0.isThunderboardDevice() || $0.name!.hasPrefix("Blinky") }
        
        case .typeMotion:
            self.presentThunderboardDeviceSelection(app: app, animated: true) { $0.isThunderboardDevice() }
            
        case .typeEnvironment:
            self.presentThunderboardDeviceSelection(app: app, animated: true) { $0.isThunderboardDevice() }
            
        case .typeThroughput:
            peripheralManager = SILThroughputPeripheralManager()
            
            weak var weakSelf = self
            peripheralManagerSubscription = peripheralManager.state.observe({ state in
                guard let weakSelf = weakSelf else { return }
                switch state {
                case .poweredOn:
                    weakSelf.peripheralManager.startAdveritising()
                    weakSelf.peripheralManagerSubscription?.invalidate()
                    
                default:
                    break
                }
            })
            self.disposeBag.add(token: peripheralManagerSubscription!)
            
            self.presentDeviceSelectionViewController(app: app, animated: true) { $0!.advertisedLocalName == "Throughput Test" }
            
        case .typeRangeTest:
            self.showRangeTest(app: app, animated: true)

        case .iopTest:
            self.presentDeviceSelectionViewController(app: app, shouldConnectWithPeripheral: false, animated: true) { $0!.advertisedLocalName?.contains("IOP") ?? false }
            
        case .typeWifiCommissioning:
            self.presentDeviceSelectionViewController(app: app, animated: true) { $0!.advertisedLocalName == "BLE_CONFIGURATOR" }
            
        default:
            return
        }
    }

    private func infoButtonTapped(recognizer: UITapGestureRecognizer) {
        self.presentAppSelectionInfoViewController(animated: true)
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.appsArray.count
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: String(describing: SILAppSelectionCollectionViewCell.self), for: indexPath) as? SILAppSelectionCollectionViewCell
        let app = self.appsArray[indexPath.row]
        cell?.setFieldsIn(app)
        return cell!
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let cellsInRow = 2.0
        let minimumLineSpacing = 16.0
        let width = floor(Double((self.appCollectionView.frame.size.width - self.appCollectionView.contentInset.left - self.appCollectionView.contentInset.right - self.appCollectionView.alignmentRectInsets.left - self.appCollectionView.alignmentRectInsets.right)) / cellsInRow) - minimumLineSpacing
        
        let height = 168.0
        
        return CGSize(width: width, height: height)
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        collectionView.deselectItem(at: indexPath, animated: true)
        let app = self.appsArray[indexPath.row]
        self.didSelectApp(app: app)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return .zero
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return 8.0
    }
    
    func deviceSelectionViewController(_ viewController: SILDeviceSelectionViewController!, didSelect peripheral: SILDiscoveredPeripheral!) {
        self.devicePopoverController?.dismissPopover(animated: true) { [self] in
            self.devicePopoverController = nil
            let appType = viewController.viewModel.app.appType
            
            switch appType {
            case .typeHealthThermometer:
                guard let peripheral = peripheral.peripheral else { return }
                self.runHealthThermometer(viewController: viewController, peripheral: peripheral)
            
            case .typeConnectedLighting:
                if let connectedLightingController = UIStoryboard(name: "SILAppTypeConnectedLighting", bundle: nil).instantiateInitialViewController() as? SILConnectedLightingViewController {
                    connectedLightingController.centralManager = viewController.centralManager
                    connectedLightingController.connectedPeripheral = peripheral.peripheral
                    self.navigationController?.pushViewController(connectedLightingController, animated: true)
                }
 
            case .typeThroughput:
                if let throughputController = UIStoryboard(name: "SILAppTypeThroughput", bundle: nil).instantiateInitialViewController() as? SILThroughputViewController {
                    throughputController.peripheralManager = peripheralManager
                    throughputController.centralManager = viewController.centralManager
                    throughputController.connectedPeripheral = peripheral.peripheral
                    self.navigationController?.pushViewController(throughputController, animated: true)
                }
                
            case .typeWifiCommissioning:
                if let wifiCommissioningController = UIStoryboard(name: "SILAppTypeWifiCommissioning", bundle: nil).instantiateInitialViewController() as? SILWifiCommissioningViewController {
                    wifiCommissioningController.centralManager = viewController.centralManager
                    wifiCommissioningController.connectedPeripheral = peripheral.peripheral
                    self.navigationController?.pushViewController(wifiCommissioningController, animated: true)
                }
                
            case .iopTest:
                let storyboard = UIStoryboard(name: "SILIOPTest", bundle: nil)
                if let iopVC = storyboard.instantiateInitialViewController() as? SILIOPTesterViewController {
                    iopVC.deviceNameToSearch = peripheral.advertisedLocalName
                    self.navigationController?.pushViewController(iopVC, animated: true)

                }
                
            default:
                break
            }
        }
    }
    
    func didDismissDeviceSelectionViewController() {
        if let peripheralManager = peripheralManager {
            peripheralManager.stopAdvertising()
        }
        
        self.devicePopoverController?.dismissPopover(animated: true)
    }

    func runHealthThermometer(viewController: SILDeviceSelectionViewController, peripheral: CBPeripheral) {
        let storyboard = UIStoryboard(name: "SILAppTypeHealthThermometer", bundle: nil)
        let controller = storyboard.instantiateInitialViewController()
        if let healthThermometerController = controller as? SILHealthThermometerAppViewController {
            healthThermometerController.centralManager = viewController.centralManager
            healthThermometerController.app = viewController.viewModel.app
            healthThermometerController.connectedPeripheral = peripheral
            self.navigationController?.pushViewController(healthThermometerController, animated: true)
        }
    }

    func presentAppSelectionInfoViewController(animated: Bool) {
        let infoViewController = SILAppSelectionInfoViewController()
        infoViewController.delegate = self
        self.devicePopoverController = WYPopoverController.sil_presentCenterPopover(withContentViewController: infoViewController, presenting: self, delegate: self, animated: true)
    }
    
    func didFinishInfo(with infoViewController: SILAppSelectionInfoViewController!) {
        self.devicePopoverController?.dismissPopover(animated: true) {
            self.devicePopoverController = nil
        }
    }
    
    func popoverControllerDidDismissPopover(_ popoverController: WYPopoverController!) {
        self.devicePopoverController?.dismissPopover(animated: true)
        self.devicePopoverController = nil
    }
    
    @objc func setIsDisconnectedIntentionallyFlag() {
        self.isDisconnectedIntentionally = true
    }

    func showThermometerPopover() {
        let apps = SILApp.demoApps() as! [SILApp]
        self.presentDeviceSelectionViewController(app: apps[0], animated: true)
        self.isDisconnectedIntentionally = false
    }
    
    func showIOPTestList(text: String) {
        let storyboard = UIStoryboard(name: "SILIOPTest", bundle: nil)
        if let iopVC = storyboard.instantiateInitialViewController() as? SILIOPTesterViewController {
            iopVC.deviceNameToSearch = text
            self.navigationController?.pushViewController(iopVC, animated: true)
        }
    }
    
    func didTappedOKButton(deviceName text: String, bluetoothState: Bool) {
        self.devicePopoverController?.dismissPopover(animated: true)
        self.devicePopoverController = nil
        if bluetoothState {
            self.showIOPTestList(text: text)
        } else {
            let bluetoothDisabledAlert = SILBluetoothDisabledAlert.interoperabilityTest
            self.alertWithOKButton(title: bluetoothDisabledAlert.title,
                                   message: bluetoothDisabledAlert.message,
                                   completion: nil)
        }
    }
    
    func didTappedCancelButton() {
        self.devicePopoverController?.dismissPopover(animated: true)
        self.devicePopoverController = nil
    }
    
    // MARK: SILThunderboardDeviceSelectionViewControllerDelegate
    
    func deviceSelectionViewControllerDidFinishThunderboardDeviceConfiguration(connection: DemoConnection, deviceConnector: DeviceConnection, appType: SILAppType) {
        SVProgressHUD.dismiss()
        self.devicePopoverController?.dismissPopover(animated: true) {
            switch appType {
            case .typeMotion:
                if let motionConnection = connection as? MotionDemoConnection, motionConnection.device.model == .sense || motionConnection.device.model == .bobcat {
                    self.displayMotion(connection: motionConnection, deviceConnector: deviceConnector)
                }
            case .typeEnvironment:
                if let environmentConnection = connection as? EnvironmentDemoConnection {
                    self.displayEnvironment(connection: environmentConnection, deviceConnector: deviceConnector)
                }
            case .typeBlinky:
                if let ioConnection = connection as? IoDemoConnection {
                    self.displayIO(connection: ioConnection, deviceConnector: deviceConnector)
                }
            default:
                return
            }
        }
    }
    
    func deviceSelectionViewControllerDidConnectWithBlinkyDevice(device: Device, deviceConnector: DeviceConnection, isThunderboard: Bool) {
        self.devicePopoverController?.dismissPopover(animated: true) {
            self.displayBlinky(device: device, deviceConnector: deviceConnector, shouldDisplayPower: isThunderboard)
        }
    }
    
    private func displayBlinky(device: Device, deviceConnector: DeviceConnection, shouldDisplayPower: Bool) {
        if let blinkyController = UIStoryboard(name: "SILAppTypeBlinky", bundle: nil).instantiateInitialViewController() as? SILAppTypeBlinkyViewController, let device = device as? BleDevice {
            blinkyController.deviceConnector = deviceConnector
            blinkyController.connectedPeripheral = device.cbPeripheral
            blinkyController.deviceName = device.name!
            if shouldDisplayPower {
                blinkyController.addConnectedDeviceBar(bottomNotchHeight: 0.0)
            }
            self.navigationController?.pushViewController(blinkyController, animated: true)
        }
    }
    
    private func displayMotion(connection: MotionDemoConnection, deviceConnector: DeviceConnection) {
        if let demoViewController = UIStoryboard(name: "MotionSenseBoardDemoViewController", bundle: nil).instantiateViewController(withIdentifier: "MotionSenseBoardDemoViewController") as? MotionSenseBoardDemoViewController {
            
            let interaction = MotionDemoInteraction(output: demoViewController, demoConnection: connection)
            demoViewController.interaction = interaction
            demoViewController.deviceConnector = deviceConnector
            demoViewController.deviceModelName = connection.device.modelName
            demoViewController.addConnectedDeviceBar(bottomNotchHeight: 0.0)
            connection.device.connectedDelegate = demoViewController
            self.navigationController?.pushViewController(demoViewController, animated: true)
        }
    }
    
    private func displayEnvironment(connection: EnvironmentDemoConnection, deviceConnector: DeviceConnection) {
        if let demoViewController = UIStoryboard(name: "EnvironmentDemoViewController", bundle: nil).instantiateViewController(withIdentifier: "EnvironmentDemoViewController") as? EnvironmentDemoViewController {
            let interaction = EnvironmentDemoInteraction(output: demoViewController, demoConnection: connection)
            demoViewController.interaction = interaction
            demoViewController.deviceConnector = deviceConnector
            demoViewController.addConnectedDeviceBar(bottomNotchHeight: 0.0)
            connection.device.connectedDelegate = demoViewController
            self.navigationController?.pushViewController(demoViewController, animated: true)
            
        }
    }
    
    private func displayIO(connection: IoDemoConnection, deviceConnector: DeviceConnection) {
        if let demoViewController = UIStoryboard(name: "IoDemoViewController", bundle: nil).instantiateViewController(withIdentifier: "IoDemoViewController") as? IoDemoViewController {
            
            let interaction = IoDemoInteraction(output: demoViewController, demoConnection: connection)
            demoViewController.interaction = interaction
            
            demoViewController.deviceConnector = deviceConnector
            demoViewController.addConnectedDeviceBar(bottomNotchHeight: 0.0)
            connection.device.connectedDelegate = demoViewController
            self.navigationController?.pushViewController(demoViewController, animated: true)
        }
    }
}
